File: server/src/Makc2021.Layer1/Query/Handlers/QueryWithOutputHandler.cs
// Copyright (c) 2021 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

using System;
using System.Collections.Generic;
using Microsoft.Extensions.Logging;

namespace Makc2021.Layer1.Query.Handlers
{
    /// <summary>
    /// Обработчик запроса с выходными данными.
    /// </summary>
    /// <typeparam name="TQueryOutput">Тип выходных данных запроса.</typeparam>    
    public class QueryWithOutputHandler<TQueryOutput> : QueryHandler, IQueryWithOutputHandler<TQueryOutput>
    {
        #region Properties

        /// <summary>
        /// Функция преобразования вывода запроса.
        /// </summary>
        protected Func<TQueryOutput, TQueryOutput> FunctionToTransformQueryOutput { get; set; }

        /// <summary>
        /// Функция получения сообщений об успехах.
        /// </summary>
        protected Func<TQueryOutput, IEnumerable<string>> FunctionToGetSuccessMessages { get; set; }

        /// <summary>
        /// Функция получения сообщений о предупреждениях.
        /// </summary>
        protected Func<TQueryOutput, IEnumerable<string>> FunctionToGetWarningMessages { get; set; }

        /// <inheritdoc/>
        public QueryResultWithOutput<TQueryOutput> QueryResult { get; private set; }

        #endregion Properties

        #region Constructors

        /// <inheritdoc/>
        public QueryWithOutputHandler(string queryName, IQueryResource appQueryResource, ILogger extLogger)
            : base(queryName, appQueryResource, extLogger)
        {
        }

        #endregion Constructors

        #region Public methods

        /// <inheritdoc/>
        public void OnStart(string queryCode = null)
        {
            DoOnStart(queryCode);
        }

        /// <inheritdoc/>
        public void OnSuccess(TQueryOutput queryOutput)
        {
            InitQueryResult(true);

            if (FunctionToTransformQueryOutput != null)
            {
                queryOutput = FunctionToTransformQueryOutput.Invoke(queryOutput);
            }

            QueryResult.Output = queryOutput;

            Func<IEnumerable<string>> functionToGetSuccessMessages = null;

            if (FunctionToGetSuccessMessages != null)
            {
                functionToGetSuccessMessages = () => FunctionToGetSuccessMessages.Invoke(queryOutput);
            }

            Func<IEnumerable<string>> functionToGetWarningMessages = null;

            if (FunctionToGetWarningMessages != null)
            {
                functionToGetWarningMessages = () => FunctionToGetWarningMessages.Invoke(queryOutput);
            }

            DoOnSuccess(functionToGetSuccessMessages, functionToGetWarningMessages);
        }

        /// <inheritdoc/>
        public void OnSuccess(QueryResultWithOutput<TQueryOutput> queryResult)
        {
            QueryResult = queryResult;

            DoOnSuccess(null, null);
        }

        #endregion Public methods

        #region Protected methods

        /// <inheritdoc/>
        protected sealed override object GetQueryInput()
        {
            return null;
        }

        /// <inheritdoc/>
        protected sealed override QueryResult GetQueryResult()
        {
            return QueryResult;
        }

        /// <inheritdoc/>
        protected sealed override void InitQueryResult(bool isOk)
        {
            QueryResult = new QueryResultWithOutput<TQueryOutput>
            {
                IsOk = isOk,
                QueryCode = QueryCode
            };
        }

        #endregion Protected methods
    }
}

File: client/makc2021-react-redux-ts/src/Layer1/Query/Handlers/QueryWithOutputHandler.ts
// Copyright (c) 2021 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

import { QueryHandler } from '../QueryHandler';
import { QueryResult } from '../QueryResult';
import {
  createQueryResultWithOutput,
  QueryResultWithOutput
} from '../QueryResultWithOutput';

/**
 * Обработчик запроса с входными и выходными данными.
 * @template TQueryOutput Тип выходных данных запроса.
 */
export abstract class QueryWithInputAndOutputHandler<
  TQueryOutput
> extends QueryHandler {
  /**
   * Результат выполнения запроса.
   */
  private _queryResult?: QueryResultWithOutput<TQueryOutput>;

  /**
   * Результат выполнения запроса.
   */
  get queryResult() {
    return this._queryResult as QueryResultWithOutput<TQueryOutput>;
  }

  /**
   * Обработать начало запроса.
   * @param queryCode Код запроса.
   */
  onStart(queryCode?: string) {
    this.doOnStart(queryCode);
  }

  /**
   * Обработать успешное выполнение запроса.
   * @param queryResult Результат выполнения запроса.
   */
  onSuccess(queryResult: QueryResultWithOutput<TQueryOutput>) {
    this._queryResult = queryResult;

    this.doOnSuccess();
  }

  /** @inheritdoc */
  protected getQueryInput() {
    return null;
  }

  /** @inheritdoc */
  protected getQueryResult() {
    return this.queryResult as QueryResult;
  }

  /** @inheritdoc */
  protected resetQueryResult() {
    this._queryResult = createQueryResultWithOutput<TQueryOutput>(
      this.queryCode
    );
  }
}
